### PR TITLE
yuzu-{ea,mainline}: {2557,953} -> {2690,992}

### PR DESCRIPTION
**Pull request by IvarWithoutBones**
###### Description of changes
This adds support for their in-app compatibility list, which previously did not work: 
image

The compatibility list does not provide any stable URLs for downloads, so we have to include it in nixpkgs and refetch it every time the program gets updated.

I've also added an update script so that r-ryantm can update this, which is always nice. When running the script manually it updates both branches, and when running it from `passthru.updateScript` it only does so for the branch from the derivation.

We're now running unit tests as well, and using SDL from nixpkgs :)

The only thing left to do is replacing the icons, as they have an unfree license. Because of that yuzu won't get build by hydra, and compiling it takes a while. Not a great user experience.

cc @sbruder @JoshuaFern 

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and “core” functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/emulators/yuzu/default.nix [hunk 1]
```diff
@@ -1,28 +1,40 @@
...
-    branchName = branch;
+    version = "2585";
+
     src = fetchFromGitHub {
       owner = "pineappleEA";
       repo = "pineapple-src";
       rev = "EA-${version}";
-      sha256 = "013xxgyn8y5fv0xbrm0zfl9xmi0gx4hpflrbjskg1hcvb2bjqyvj";
+      sha256 = "sha256-TTfjJBqbIRsHa8draDpqSvozGbgA5/SPSSHV7IESLEg=";
+      fetchSubmodules = true;
```
Comment (IvarWithoutBones) [inline pkgs/applications/emulators/yuzu/default.nix:35]: For whatever reason `nix-prefetch-git` generates a wrong checksum when not using `fetchSubmodules`/`--fetch-submodules`. It doesn't actually have any submodules, this is just done for the update script.

**Review by sbruder — COMMENTED** [review]
This is very nice! Thanks for all your effort.

Just a few things I noticed:

SDL being used from nixpkgs makes things less cluttered and also improves compile times. However we should keep an eye on upstream bumping the SDL requirement at which point we have to also upgrade SDL or switch to the bundled SDL again. This isn’t a real problem right now, just a heads up for future updates.

The compatibility list is quite large and since it will probably be updated with every update, this will lead to huge changes over time. I don’t know if this causes any problems or is wanted in nixpkgs in general. Sadly, including this is the only option currently possible (and the compatibility list is quite nice, so it should be included). Maybe we should ask upstream if they are willing to provide snapshots of the compatibility list?

Replacing the icons is also something that should be done upstream if possible. There is [a list](https://github.com/yuzu-emu/yuzu/blob/master/dist/license.md) of non-free icons in the yuzu source code. Many of them were just borrowed from citra (see [their list](https://github.com/citra-emu/citra/blob/master/dist/license.md)), so this also affects citra (which currently only has `gpl2` set as its license). It [has been noted](https://github.com/citra-emu/citra/pull/5093#issuecomment-604676784) that this is a problem for distributions, though this has gone unnoticed.

While the update script works when used standalone, it does not work with the maintainer update script (like with `nix-shell maintainers/scripts/update.nix --argstr package yuzu-mainline`). Am I doing this wrong or is there something that the maintainer update script expects, that this update script does not implement?

File: pkgs/applications/emulators/yuzu/update.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+        echo "Compatibility list: updated to $TODAY"
+        echo "${NEW_COMPAT_LIST}" > ./compatibility-list.json
+
+        sed -i -e "s/last updated .*/last updated $TODAY./" ./default.nix
+    fi
+}
+
+updateMainline() {
+    pushd ../../../.. >/dev/null
```
Comment (SuperSandro2000) [inline pkgs/applications/emulators/yuzu/update.sh:38]: I think we need to add pushd and popd to the nix-shell.

File: pkgs/applications/emulators/yuzu/generic.nix [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+  # This must be done after cmake finishes as it overwrites the file
+  postConfigure = ''
+    cp ${compat-list} ./dist/compatibility_list/compatibility_list.json
+  '';
+
+  # Fix vulkan detection
+  postFixup = ''
+    wrapProgram $out/bin/yuzu --prefix LD_LIBRARY_PATH : ${vulkan-loader}/lib
+    wrapProgram $out/bin/yuzu-cmd --prefix LD_LIBRARY_PATH : ${vulkan-loader}/lib
```
Comment (SuperSandro2000) [inline pkgs/applications/emulators/yuzu/generic.nix:124]:
```suggestion
    for bin in $out/bin/yuzu $out/bin/yuzu-cmd; do
      wrapProgram $file --prefix LD_LIBRARY_PATH : ${vulkan-loader}/lib
   done
```
just an idea

File: pkgs/applications/emulators/yuzu/generic.nix [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    # see https://github.com/NixOS/nixpkgs/issues/114044, setting this through cmakeFlags does not work.
+    cmakeFlagsArray+=(
+      "-DTITLE_BAR_FORMAT_IDLE=yuzu ${branch} ${version}"
+      "-DTITLE_BAR_FORMAT_RUNNING=yuzu ${branch} ${version} | {3}"
+    )
+  '';
+
+  # This must be done after cmake finishes as it overwrites the file
+  postConfigure = ''
+    cp ${compat-list} ./dist/compatibility_list/compatibility_list.json
```
Comment (SuperSandro2000) [inline pkgs/applications/emulators/yuzu/generic.nix:118]: Could we symlink it?

File: pkgs/applications/emulators/yuzu/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/applications/emulators/yuzu/default.nix:35]: Did you try if only leaveDotGit works?

Comment (IvarWithoutBones) [reply pkgs/applications/emulators/yuzu/default.nix:35]: I haven't, as I had [some issues](https://github.com/NixOS/nixpkgs/pull/84117#discussion_r411703390) with it when initially adding this package. See also https://github.com/NixOS/nixpkgs/issues/8567.

File: pkgs/applications/emulators/yuzu/generic.nix (hunk 4 above)
Comment (IvarWithoutBones) [reply pkgs/applications/emulators/yuzu/generic.nix:118]: Yes :+1:

**Comment by sbruder** [comment]
I opened an issue upstream (https://github.com/yuzu-emu/yuzu/issues/8102) to ask about the feasibility of snapshots of the game compatibility database, though we can just include it in nixpkgs for the time being.

Also, in the time since you opened this PR, upstream [changed the license to gpl3Plus](https://github.com/yuzu-emu/yuzu/commit/284934ebfdf5e530c960cf69969172ff76f40bea), so this should also be reflected in this update.

**Comment by IvarWithoutBones** [comment]
> I opened an issue upstream (https://github.com/yuzu-emu/yuzu/issues/8102) to ask about the feasibility of snapshots of the game compatibility database, though we can just include it in nixpkgs for the time being.

Sounds good :+1: 

> Also, in the time since you opened this PR, upstream [changed the license to gpl3Plus](https://github.com/yuzu-emu/yuzu/commit/284934ebfdf5e530c960cf69969172ff76f40bea), so this should also be reflected in this update.

Done. I've also updated both brances to the latest versions again. I think there should be nothing blocking this anymore :blush:

**Review by sbruder — APPROVED** [review]
Despite the maintainer update script not working with the update script (which I don’t know how to fix and which can also be fixed in a later PR), this looks good.

File: pkgs/applications/emulators/yuzu/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/applications/emulators/yuzu/default.nix:35]: IIRC fetchSubmodules has the same issues.

Comment (IvarWithoutBones) [reply pkgs/applications/emulators/yuzu/default.nix:35]: It seems like `leaveDotGit` doesn't fix this, sadly.

**Comment by IvarWithoutBones** [comment]
I've rebased this to the latest versions again, unless anyone has any objections I'll go ahead and merge this in a few days :+1:

**Comment by Artturin** [comment]
seriously, a 1.2M file that could just be fetched from a archive site? please switch to using a archive site

```
$ fd --type f --size "+1MiB"
pkgs/applications/editors/emacs/elisp-packages/recipes-archive-melpa.json
pkgs/applications/emulators/yuzu/compatibility-list.json
pkgs/desktops/gnome/extensions/extensions.json
pkgs/development/haskell-modules/hackage-packages.nix
pkgs/development/node-packages/node-packages.nix
pkgs/development/r-modules/cran-packages.nix
pkgs/servers/monitoring/prometheus/webui/webui/package-lock.json
pkgs/tools/typesetting/tex/texlive/pkgs.nix
pkgs/top-level/all-packages.nix
```

**Comment by Artturin** [comment]
this is seriously messing up my grepping @IvarWithoutBones

**Pull request merged**